File: backend/generateInstrumentPartsAndMaster.js
const { PDFDocument } = require('pdf-lib')
const fs = require('fs')
const validFilename = require('valid-filename')

const directorySeparator = process.platform === 'win32' ? '\\' : '/'

const generateInstrumentPartsAndMaster = async (
	collectionName,
	songFoldersLocation,
	songList,
	instrumentPartsList
) => {
	const destinationDirectory = `${songFoldersLocation}${directorySeparator}${collectionName}`

	if (!validFilename(prefix)) {
		throw `"${collectionName}" is not a valid folder name. Please try something different.`
	}

	if (!fs.existsSync(songFoldersLocation)) {
		throw 'Error: The Song Folders Location does not exist.'
	}

	if (fs.existsSync(destinationDirectory)) {
		throw `Error: Cannot create destination folder.\n\nA folder named "${destinationDirectory}" already exists.`
	}

	// generate list of parts and number of respective copies based on given input
	let numCopiesOfPart = {}
	for (part of instrumentPartsList.split(/\r?\n/)) {
		// split line into part name and number of copies
		let partNameAndNumCopies = part.split('#')

		// remove spaces from part name
		partNameAndNumCopies[0] = partNameAndNumCopies[0].replace(/\s/g, '')

		// skip part if blank
		if (!partNameAndNumCopies[0]) {
			continue
		}

		// set default num copies to 1 or convert num copies from str to int
		if (partNameAndNumCopies.length < 2) {
			partNameAndNumCopies.push(1)
		} else {
			partNameAndNumCopies[1] = parseInt(partNameAndNumCopies[1])
		}

		// add part and its number of pages to dict
		numCopiesOfPart[partNameAndNumCopies[0]] = partNameAndNumCopies[1]
	}

	// create Master PDF document (contains all copies of all parts)
	let masterPdfDocument = await PDFDocument.create()

	// create PDF for each instrument
	let instrumentPdfFiles = []
	for (instrument in numCopiesOfPart) {
		const instrumentPartFilePath = `${destinationDirectory}${directorySeparator}${collectionName}-${instrument}.pdf`

		let instrumentPartPdfDocument = await PDFDocument.create()

		// get this instrument's part from each song
		for (songName of songList.split(/\r?\n/)) {
			// remove spaces from song name
			songName = songName.replace(/\s/g, '')

			// skip songName if blank
			if (!songName) {
				continue
			}

			// get instrument part PDF from this song
			const songDirectory = `${songFoldersLocation}${directorySeparator}${songName}`
			const partFile = fs.readFileSync(
				`${songDirectory}${directorySeparator}${songName}-${instrument}.pdf`
			)
			let partFileReader = await PDFDocument.load(partFile)

			// copy the pages to the instrument part
			const copiedPages = await instrumentPartPdfDocument.copyPages(
				partFileReader,
				partFileReader.getPageIndices()
			)
			for (page of copiedPages) {
				instrumentPartPdfDocument.addPage(page)
			}

			// if working with a score, special case, add a blank page after each if there is an odd number of pages in the score
			if (
				instrument.toLowerCase() === 'score' &&
				partFileReader.getPageCount() % 2 === 1
			) {
				addBlankPage(instrumentPartPdfDocument)
			}
		}

		// export instrument part PDF
		const isntrumentPartFileData = await instrumentPartPdfDocument.save()
		instrumentPdfFiles.push({
			path: instrumentPartFilePath,
			data: isntrumentPartFileData,
		})

		// add instrument part PDF to master PDF
		for (let i = 0; i < numCopiesOfPart[instrument]; i++) {
			const copiedPages = await masterPdfDocument.copyPages(
				instrumentPartPdfDocument,
				instrumentPartPdfDocument.getPageIndices()
			)
			for (page of copiedPages) {
				masterPdfDocument.addPage(page)
			}

			// if odd number of pages in instrument part, add blank page for double sided printing
			if (instrumentPartPdfDocument.getPageCount() % 2 === 1) {
				addBlankPage(masterPdfDocument)
			}
		}
	}

	// create destination directory for the new PDF files
	fs.mkdirSync(destinationDirectory)

	// export isntrument part PDF files
	for (instrumentPdfFile of instrumentPdfFiles) {
		fs.writeFileSync(instrumentPdfFile.path, instrumentPdfFile.data)
	}

	// export master PDF file
	fs.writeFileSync(
		`${destinationDirectory}/${collectionName}-MASTER.pdf`,
		await masterPdfDocument.save()
	)

	return destinationDirectory
}

const addBlankPage = (pdfDocument) => {
	const previousPageSize = pdfDocument
		.getPage(pdfDocument.getPageCount() - 1)
		.getSize()
	// add a blank page with the same dimensions as the previos page
	pdfDocument.addPage([previousPageSize.width, previousPageSize.height])
}

module.exports = { generateInstrumentPartsAndMaster }

File: backend/generate_instrument_parts_and_master.py
import sys
import os
import PyPDF2

def generate_instrument_parts_and_master(piece_list, song_folders_location, parts_list, dest_name):
	if not os.path.exists(song_folders_location):
		raise FileNotFoundError('The Song Folders Location does not exist')
	
	# create list of parts and number of respective copies
	num_copies_of_part = dict()
	for part in parts_list.splitlines():
		# split line into part name and number of copies
		part_name_and_num_copies = part.split("#")

		# remove spaces from part name
		part_name_and_num_copies[0] = part_name_and_num_copies[0].strip() \
			.replace(" ", "")

		# set default num copies to 1 or convert num copies from str to int
		if len(part_name_and_num_copies) < 2:
			part_name_and_num_copies.append(1)
		else:
			part_name_and_num_copies[1] = int(part_name_and_num_copies[1])

		# add part and its number of pages to dict
		num_copies_of_part[part_name_and_num_copies[0]] \
			= part_name_and_num_copies[1]

	# create new dir named dest_name and chdir to that dir
	os.chdir(song_folders_location)
	os.mkdir(dest_name)
	os.chdir(dest_name)

	# create PDF for each instrument
	master_writer = PyPDF2.PdfFileWriter()
	master_file_name = "{}-MASTER.pdf".format(dest_name)
	for part in num_copies_of_part.keys():
		part_file_name = "{}-{}.pdf".format(dest_name, part)
		part_writer = PyPDF2.PdfFileWriter()

		# merge parts from each piece
		for piece in piece_list.splitlines():
			# remove spaces from piece name
			piece = piece.strip().replace(" ", "")

			os.chdir("../{}".format(piece))
			# open part_files (yes, it has to be done this way...)
			part_files = []
			part_files.append(open("{}-{}.pdf".format(piece, part), "rb"))
			part_file_reader = PyPDF2.PdfFileReader(part_files[-1])
			part_writer.appendPagesFromReader(part_file_reader)

			# if working with a score, special case, add a blank page after each
			if part.upper() == "SCORE" and part_file_reader.numPages %2 == 1:
				part_writer.addBlankPage()

		# export part PDF
		os.chdir("../{}".format(dest_name))
		with open(part_file_name, "wb") as part_pdf_writer:
			part_writer.write(part_pdf_writer)

		# open part_pdf_readers (yes, it has to be done this way...)
		part_pdf_readers = []
		part_pdf_readers.append(open(part_file_name, "rb"))
		# append part PDF to master PDF
		part_file_reader = PyPDF2.PdfFileReader(part_pdf_readers[-1])
		for i in range(0, num_copies_of_part[part]):
			master_writer.appendPagesFromReader(part_file_reader)
			# if odd number of pages, add blank page for double sided printing
			if part_file_reader.numPages %2 == 1:
				master_writer.addBlankPage()

	# export master PDF
	with open(master_file_name, "wb") as master_file:
		master_writer.write(master_file)

	# close part_pdf_readers (yes, it has to be done this way...)
	for i in range(len(part_pdf_readers)):
		part_pdf_readers[i].close()

	# close part_files (yes, it has to be done this way...)
	for i in range(len(part_files)):
		part_files[i].close()

	return f'{song_folders_location}/{dest_name}'

if __name__ == "__main__":
	# read cmd line args
	piece_list = ''
	for i in range(1, len(sys.argv) - 2):
		piece_list += sys.argv[i] + '\n'

	parts_list_path = sys.argv[-2]
	with open(parts_list_path) as parts_list_file:
		parts_list = parts_list_file.read()

	dest_name = sys.argv[-1]

	generate_instrument_parts_and_master(piece_list, os.getcwd(), parts_list, dest_name)